Feat: add command loop

Dzarda7 opened the pull request:
## Description
This PR adds command loop and basic command handling. SLIP packet receive and send needs to be added. Protocol command handling, especially flashing ones needs to be implemented.

Needs to be revised later for faster communication (consider avoid using memcpy, implement double buffering).

## Testing
Only build tested

Dzarda7 commented on src/command_handler.cpp:
```diff
@@ -0,0 +1,490 @@
+    }
+
+    template<typename T>
+    void init_from(const T &params)
```
Now in C.

Dzarda7 commented:
@dobairoland @radimkarnis @jedi7 PTAL. This will need modifications later, but should be good enough for initial function.

dobairoland commented on src/command_handler.c:
```diff
@@ -0,0 +1,494 @@
+static operation_state_t s_flash_state = {0};
+static operation_state_t s_memory_state = {0};
+
+static void handle_sync(void);
```
(optional) You might consider naming static functions in a way that it will be clear to reader that there are local, e.g. `s_sync` - dropping "handle" saves space and the line length can be more effective used and reduce the need to splitting up lines. Or an abbreviation of the module, e.g. "ch_"

Dzarda7 commented on src/command_handler.c:
```diff
@@ -0,0 +1,494 @@
+    const read_reg_params_t *params = (const read_reg_params_t*)data;
+    const uint32_t value = read_reg(params->addr);
+
+    send_success_response(ESP_READ_REG, value, NULL, 0);
```
Thanks for the suggestion, I thought about that, but I have some things I want to test in mind and this gives me a bit more flexibility for now. Will remember that definitely.

Dzarda7 commented on src/command_handler.c:
```diff
@@ -0,0 +1,494 @@
+    response_buffer[sizeof(packet) + 1] = (uint8_t)error;
+
+    if (data_len > 0 && data != NULL) {
+        memcpy(response_buffer, data, data_len);
```
nice catch

Dzarda7 commented on src/command_handler.c:
```diff
@@ -0,0 +1,494 @@
+{
+    // TODO: Consider verifying data length with known command size
+    if (!s_memory_state.in_progress) {
+        send_error_response(ESP_MEM_DATA, NOT_IN_FLASH_MODE);
```
Keeping for compatibility with esptool

dobairoland commented on src/command_handler.c:
```diff
@@ -0,0 +1,474 @@
+    }
+
+    send_success_response(ESP_WRITE_REG, 0, NULL, 0);
+}
```
This is also micro-optimization. Using unnecessary post increment used to be less effective because it stored the previous value before incrementing and then throw it away.
 
Pre-increment should be used by default:
```c
++s_flash_state.blocks_written;
```

And post-increment only if:
```
i_want_the_old_value = s_flash_state.blocks_written++;
```

Dzarda7 commented on src/command_handler.cpp:
```diff
@@ -0,0 +1,490 @@
+    }
+};
+
+static OperationState s_flash_state;
```
In C now. Thanks.

erhankur commented on src/command_handler.c:
```diff
@@ -0,0 +1,474 @@
+        data_len = MAX_RESPONSE_DATA_SIZE;
+    }
+
+    const size_t combined_len = sizeof(response_status_t) + sizeof(error_code_t) + data_len;
```
combined_len = data_len + 8

but below you are copying 2 bytes for response_status_t and error_code_t.   

combined_len should be data_len + 2  and uint16_t to avoid casting ?

erhankur replied:
Thanks. This is a good example why we should avoid casting and fix the types instead. `enum class : uint8_t` would be suitable for this scenario.  But we are in c.

Dzarda7 replied:
Yes, that is where it originated probably :smile:.

erhankur commented on src/commands.h:
```diff
@@ -0,0 +1,154 @@
+ * @brief ESP command error codes
+ * Used as the error_msg byte when status is FAIL
+ */
+typedef enum {
```
Might be good to add a comment that max value must be 0xFF to avoid bug in the response message.  same as response_status_t

Dzarda7 replied:
Done, thanks.

Dzarda7 commented on src/command_handler.c:
```diff
@@ -0,0 +1,474 @@
+
+        case ESP_ERASE_REGION:
+            s_erase_region(data, packet->size);
+            break;
```
I believe I did if you mean all post ++

Dzarda7 commented on src/command_handler.c:
```diff
@@ -0,0 +1,474 @@
+inline uint32_t calculate_checksum(const void* data, size_t len)
+{
+    uint32_t checksum = 0xEF;
+    const uint8_t* bytes = (const uint8_t*)data;
```
Sorry, I missed this one. Thanks for noticing.

Dzarda7 commented on src/command_handler.c:
```diff
@@ -0,0 +1,494 @@
+#include "commands.h"
+#include "command_handler.h"
+
+#define MAX_COMMAND_SIZE (sizeof(common_command_t) + 0x4000)
```
Yes, added in the code, it is temporary because of esptool

erhankur commented on src/command_handler.c:
```diff
@@ -0,0 +1,475 @@
+        data_len = MAX_RESPONSE_DATA_SIZE;
+    }
+
+    const uint16_t combined_len = (uint16_t)(data_len + 2); // 2 bytes for status and error
```
should combined_len include the packet len?
```c
combined_len =  sizeof(packet) + data_len + 2;
```

erhankur reviewed (approved):
One last question and LGTM. Thanks

Dzarda7 commented:
Thanks @erhankur and @dobairoland for deep review. I believe I addressed all the issues at least with a comment. This is just a first iteration to have it working. When reception is working, I plan to play a bit with ringbuffer and partial command processing.

radimkarnis commented on src/command_handler.c:
```diff
@@ -0,0 +1,475 @@
+    bool in_progress;
+} operation_state_t;
+
+// 0x4000 is the maximum data size sent by esptool, so keeping for, so keeping the compatibility
```
`so keeping for compatibility`

Dzarda7 replied:
Thanks, fixed.

radimkarnis reviewed (approved):
Other than the above-mentioned suggestions, this LGTM! Thanks.

jedi7 commented on src/command_handler.c:
```diff
@@ -0,0 +1,475 @@
+    slip_send_frame(response_buffer, sizeof(packet) + resp_data_len);
+}
+
+void command_handler_loop(void)
```
I will prefer to move this loop into main and simplify it. See below.

jedi7 commented on src/command_handler.c:
```diff
@@ -0,0 +1,475 @@
+    uint8_t data_buffer[MAX_COMMAND_SIZE];
+
+    for (;;) {
+        size_t len = 0; // TODO: Receive SLIP frame
```
what about to move these things into a function receive(buffer) and process(buffer) ?
And keep these functions here and main loop can be moved to main.c
It will be easier to test that.

Dzarda7 replied:
Done, thanks. In the future it might make sense to separate it from SLIP completely and return response from the function.

jedi7 commented on src/commands.h:
```diff
@@ -0,0 +1,157 @@
+ * Should be max 0xFF to be compatible with protocol
+ */
+typedef enum {
+    ESP_SYNC = 0x08,
```
please move the esp_sync after esp_mem_end to keep the continuity

jedi7 commented on src/command_handler.c:
```diff
@@ -0,0 +1,475 @@
+#include "commands.h"
+#include "command_handler.h"
+
+typedef struct __attribute__((packed))
```
todo: we need to do something with the attribute packed and do proper deserialization

Dzarda7 replied:
Yes, we should consider it, it will even improve type checking.

jedi7 commented on src/command_handler.c:
```diff
@@ -0,0 +1,475 @@
+typedef struct __attribute__((packed))
+{
+    uint8_t direction;      /* 0x00 for request, 0x01 for response */
+    esp_command_t command;     /* Command ID */
```
this enum can have different base type. Imho it will depend on compiler.

jedi7 commented on src/commands.h:
```diff
@@ -0,0 +1,157 @@
+ * @brief ESPTool flasher stub commands
+ * Should be max 0xFF to be compatible with protocol
+ */
+typedef enum {
```
can we enforce base type of the enum?

jedi7 commented on src/command_handler.h:
```diff
@@ -0,0 +1,39 @@
+ * @param len Length of the buffer
+ */
+
+command_handler_status_t handle_command(uint8_t buffer[], size_t len);
```
I'm curious, why buffer[] and not *buffer?

fhrbata commented on src/command_handler.c:
```diff
@@ -0,0 +1,464 @@
+{
+    (void)len; // TODO: Consider verifying data length with known command size
+
+    const flash_begin_params_t *params = data;
```
Dummy question,  what guarantees that the `data` buffer is properly [aligned](https://docs.espressif.com/projects/esp-idf/en/stable/esp32/api-guides/fatal-errors.html#loadstorealignment)?

Dzarda7 replied:
Added static and aligned to 4 bytes, thanks for noticing.

fhrbata commented on src/command_handler.c:
```diff
@@ -0,0 +1,464 @@
+
+static void s_flash_begin(const void* data, uint16_t len)
+{
+    (void)len; // TODO: Consider verifying data length with known command size
```
IMHO this is a must.

fhrbata commented on src/commands.h:
```diff
@@ -0,0 +1,157 @@
+
+// Command parameter structures
+
+typedef struct __attribute__((packed))
```
Is the `packed`  really needed? All the structs seem to be naturally aligned. Maybe you can check with [pahole](https://linux.die.net/man/1/pahole). IMHO more robust approach would be to parse the input buffer, not casting it.

fhrbata commented:
@Dzarda7 feel free to ignore/resolve my dummy questions 😀. Thanks

Dzarda7 commented:
@dobairoland @radimkarnis @jedi7 @fhrbata PTAL. I removed casting and structures, added size check for each command and used static buffer instead of stack buffer. Thanks for all suggestions.

I might have missed something, so feel free to suggest improvements. Also feel free to suggest bigger changes, all of it can be addressed later, but can be used to improve the final code. Main aim of this is to have something working with esptool, later when we focus on improvements this is flexible enough to test possible speedups (multiple buffers, partial command handling etc.), after these are implemented, I see a few possible code quality improvements.

dobairoland commented on src/command_handler.c:
```diff
@@ -0,0 +1,627 @@
+
+    uint32_t flag;
+
+    memcpy(&flag, buffer, sizeof(uint32_t));
```
Now there a bunch of memcopies of length 4 .

A question for future optimization: How does the compiled machine code looks like? Do we want rather a for cycle iterating between 0 and 3, or have 4 direct assignment (without 4 additional incrementations, substractions and jumps). Will the compiler optimize these in a reasonable way?

Dzarda7 replied:
I redid it, I missed that most of the params are uint32_t, so it looks nicer like this IMO, thanks.

dobairoland commented on src/commands.h:
```diff
@@ -0,0 +1,85 @@
+ * @brief Response status - SUCCESS or FAIL
+ * Should be max 0xFF to be compatible with protocol
+ */
+enum {
```
I think these are better as macros in C.

Dzarda7 replied:
Done, thanks, thought about that.

dobairoland commented on src/commands.h:
```diff
@@ -0,0 +1,85 @@
+ * Used as the error_msg byte when status is FAIL
+ * Should be max 0xFF to be compatible with protocol
+ */
+enum {
```
Likewise. I don't see a reason to put them into an anonymous enum if all the values are set. You only get the uncertainty about the datatype.

dobairoland commented on src/main.c:
```diff
@@ -27,7 +30,13 @@ void esp_main(void)
     slip_send_frame(&greeting, sizeof(greeting));
 
-    // TODO: Implement command loop
+    while (1) {
```
What happened to `for(;;)`?

Dzarda7 replied:
Fixed, thanks, I thought main loop convention is while(1)

dobairoland reviewed (approved):
LGTM. Feel free to merge it and address comments in a future PR.

jedi7 commented on src/command_handler.c:
```diff
@@ -0,0 +1,627 @@
+
+static void s_flash_begin(const uint8_t* buffer, uint16_t size)
+{
+    const uint8_t expected_size = 4 * sizeof(uint32_t);
```
nit: it will be nice to explain this magic number 4 (comment or constant)

jedi7 commented on src/command_handler.c:
```diff
@@ -0,0 +1,627 @@
+
+    uint32_t total_size, num_blocks, block_size, offset;
+
+    memcpy(&total_size, buffer, sizeof(uint32_t));
```
nit: is always preferred to use sizeof a variable , than a type. Then you can easily change the type size and nothing will break.
Also the buffer offset can be done by sizeof or increasing the buffer pointer

jedi7 commented on src/command_handler.c:
```diff
@@ -0,0 +1,627 @@
+    memcpy(&block_size, buffer + 8, sizeof(uint32_t));
+    memcpy(&offset, buffer + 12, sizeof(uint32_t));
+
+    s_flash_state.total_size = total_size;
```
Imho, is ok to memcpy directly to the structure attributes to avoid multiple copying

jedi7 commented on src/command_handler.c:
```diff
@@ -0,0 +1,627 @@
+    const uint16_t command_count = (uint16_t)(size / single_reg_size);
+
+    for (uint16_t i = 0; i < command_count; i++) {
+        const uint8_t* reg_bytes = buffer + (i * single_reg_size);
+        uint32_t addr, value, mask, delay_us;
+
+        memcpy(&addr, reg_bytes, sizeof(uint32_t));
+        memcpy(&value, reg_bytes + 4, sizeof(uint32_t));
+        memcpy(&mask, reg_bytes + 8, sizeof(uint32_t));
+        memcpy(&delay_us, reg_bytes + 12, sizeof(uint32_t));
```
the copilot have good point

jedi7 commented on src/command_handler.c:
```diff
@@ -0,0 +1,627 @@
+    uint8_t direction_byte = DIRECTION_RESPONSE;
+
+    memcpy(response_buffer, &direction_byte, sizeof(uint8_t));
+    memcpy(response_buffer + 1, &command, sizeof(uint8_t));
```
nit: again sizeof variables, also on other places :)

jedi7 commented on src/command_handler.c:
```diff
@@ -0,0 +1,627 @@
+    }
+
+    const uint16_t resp_data_size = (uint16_t)(data_size + RESPONSE_STATUS_SIZE);
+    const uint16_t total_frame_size = (uint16_t)(HEADER_SIZE + data_size + RESPONSE_STATUS_SIZE);
```
(HEADER_SIZE + resp_data_size)

jedi7 commented on src/command_handler.h:
```diff
@@ -0,0 +1,34 @@
+#endif
+
+// 0x4000 is the maximum data size sent by esptool, so keeping for the compatibility
+#define ESPTOOL_MAX_DATA_SIZE 0x4000U
```
do we need these defines in "public" api? I think it will be ok to hide them to the c file

Dzarda7 commented:
I hope I addressed all @jedi7 @dobairoland, thanks for the great points.

dobairoland commented on src/commands.h:
```diff
@@ -0,0 +1,105 @@
+ * @brief ESPTool flasher stub commands
+ * Should be max 0xFF to be compatible with protocol
+ */
+enum {
```
This is another enum related to my previous comment.

Dzarda7 replied:
Fixed, thanks.

radimkarnis commented:
Thanks for resolving the suggestions! LGTM.

Dzarda7 commented:
Thanks all for many great suggestions.
